### PR TITLE
Deprecate whitebox and corresponding verboserunner and junitfailureha…

### LINKED ISSUE #489: Remove Whitebox class
The original question stems from #422 where additional changes were requested. This class seems to stimulate bad testing practices. Mockito only uses this class in [JUnitFailureHacker](https://github.com/mockito/mockito/blob/196ff979da156caa07e19f57e4849637d8bede1a/src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java) which consequently is only used in [VerboseMockitoJUnitRunner](https://github.com/mockito/mockito/blob/196ff979da156caa07e19f57e4849637d8bede1a/src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java). Given the nature of this class and only 1 usage in the library, I think we should remove it to prevent users from obtaining bad testing habits.

### PR DESCRIPTION
**Pull request by TimvdLippe**
A quick google showed that there are almost zero usages of this runner. Moreover it was relying on a very hackish implementation. I think we can do a better job by creating a new runner with JUnit 5 as part of #390. Therefore the intent is to deprecate these classes and remove them in Mockito 3.0.

Fixes #489

### CODE REVIEW

**Comment by bric3** [comment]
Don't remove this runner, a few people were relying on this. It may not show up on open source project.

In my opinion `Whitebox` should be removed though. So it could be possible to use other internal reflection utils, like `FieldGetter`

File: src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java [hunk 1]
```diff
@@ -34,7 +34,11 @@
  * Do you think it is useful or not? Drop us an email at mockito@googlegroups.com
  * <p>
  * Experimental implementation - will change in future!
+ *
+ * @deprecated as of 2.0. Runner will likely be replaced with a JUnit 5 compliant runner. This runner relied
+ * on a hacky implementation as was not used widely.
```
Comment (Xaerxess) [inline src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java]: "as" -> "and"?

Comment (bric3) [inline src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java]: Also sync with @szczepiq on this

**Comment by ChristianSchwarz** [comment]
@TimvdLippe @bric3 
What do you think about deprecating all Mockito-Runners in order to promote the uses of `MockitoRule`? 

Motivation: 
 * Mockito should promote only one API to initialize annotated fields, cause it may confuse users what to choose when (Runners vs Rule). 
 * `MockitoRule `does the same as the Runners but is more flexible cause it allows to uses other Runners e.g. for parameterized tests which is a often requested feature

**Comment by raphw** [comment]
I do not think we should deprecate them. JUnit (4) still has runners as a non-deprecated feature. We should deprecate the runners once they do the same. A lot of people still rely on old JUnit versions where runners are the only option. We can however add proper documentation that indicates that runners should be the second choice.

**Comment by ChristianSchwarz** [comment]
@raphw 
Even if I disagree at some points I accept the core-team decision. 

> A lot of people still rely on old JUnit versions where runners are the only option.

I see you point, but i think that people that update to mockito 2.x can also update there junit version.

> We can however add proper documentation that indicates that runners should be the second choice.

Good idea, here we go #531

**Comment by TimvdLippe** [comment]
The reason to deprecate this runner is because the new Mockito runner is strict and is a lot more userfriendly than this runner. To promote the strict runner, we should deprecate this one and use the strict version imo.

Regarding runners vs rules: we can state our preference, but we should keep the possibility.

**Comment by szczepiq** [comment]
I'll tidy up the Verbose runner documentation. I don't think it is useful given the work on #384 and #401.

**Comment by bric3** [comment]
OK deprecation sounds good to me, however I'd remove `Whitebox it's internal after all.

Let's just replace the methods in `JUnitFailureHacker` at [line 19](https://github.com/mockito/mockito/pull/491/files#diff-a041e60ee2d1a736352af1c34c19b1f4R19) and at [line 26](https://github.com/mockito/mockito/pull/491/files#diff-a041e60ee2d1a736352af1c34c19b1f4R26) by private methods.

File: src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java (hunk 1 above)
Comment (TimvdLippe) [inline src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java]: Documentation updated :smile:

**Comment by TimvdLippe** [comment]
Rebased and updated pull request with feedback. PTAL :smile:

File: src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java [hunk 2]
```diff
@@ -6,26 +6,73 @@
 
 import org.junit.runner.notification.Failure;
 import org.mockito.internal.exceptions.ExceptionIncludingMockitoWarnings;
-import org.mockito.internal.util.reflection.Whitebox;
 
+import java.lang.reflect.Field;
+
+@Deprecated
```
Comment (ChristianSchwarz) [inline src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java]: It should be documented what should be used instead.

Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java]: It's an internal class and is only used in the VerboseMockitoRunner. Therefore I do not think documentation is required.

File: src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java [hunk 3]
```diff
@@ -6,26 +6,73 @@
...
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to get internal state on a private field. Please report to mockito mailing list.", e);
+        }
+    }
+
+    private static void setInternalState(Object target, String field, Object value) {
+        Class<?> c = target.getClass();
+        try {
+            Field f = getFieldFromHierarchy(c, field);
+            f.setAccessible(true);
```
Comment (ChristianSchwarz) [inline src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java]: @TimvdLippe Does it make sense to uses the AccessabilityChanger here?

File: src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java [hunk 4]
```diff
@@ -34,7 +34,10 @@
  * Do you think it is useful or not? Drop us an email at mockito@googlegroups.com
  * <p>
  * Experimental implementation - will change in future!
+ *
+ * @deprecated as of 2.0. Use the new {@link org.mockito.runners.MockitoJUnitRunner.Strict} runner instead.
```
Comment (ChristianSchwarz) [inline src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java]: @TimvdLippe cosmetics: 2.0.0 is used in other places, it should be used here too

File: src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java (hunk 3 above)
Comment (TimvdLippe) [inline src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java]: It's more code with object creation, so I think this one-liner is better.

File: src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java (hunk 4 above)
Comment (TimvdLippe) [inline src/main/java/org/mockito/runners/VerboseMockitoJUnitRunner.java]: Done

File: src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java (hunk 3 above)
Comment (bric3) [inline src/main/java/org/mockito/internal/util/junit/JUnitFailureHacker.java]: And especially for deprecated code

**Comment by bric3** [comment]
Indeed the timeout test is picky, especially with openjdk 6, I restarted the job 2 times

**Comment by szczepiq** [comment]
Verbose runner and console spamming runner are now both deprecated. Let's kill them in 3.0 :)

**Pull request merged**